clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.Core.Evaluator.Models where
+
+import Control.Concurrent.Supply (Supply)
+import Control.Monad.State.Strict (State)
+import Control.DeepSeq (NFData(..))
+import Data.Bifunctor (first, second)
+import Data.Foldable (foldl')
+import Data.IntMap.Strict (IntMap)
+import qualified Data.IntMap.Strict as IntMap
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+
+import BasicTypes (InlineSpec(..))
+import SrcLoc (noSrcSpan)
+
+import Clash.Core.DataCon
+import Clash.Core.Literal
+import Clash.Core.Subst
+import Clash.Core.Term
+import Clash.Core.TyCon
+import Clash.Core.Type
+import Clash.Core.Var
+import Clash.Core.VarEnv
+import Clash.Driver.Types
+
+-- | An evaluator contains the basic functions that define partial evaluation.
+-- This consists of:
+--
+--   * evaluate, which evaluates terms to WHNF
+--   * quote, which recursively evaluates sub-terms and eta-expands to NF
```

Comment:
Why not: `evalWhnf` / `evalNf` (or with "To" interspersed)? Seems a lot more self-descriptive.

Reply:
Normally in literature it would be eval / quote, although I see your point.

What about changing `evaluate` to `evaluateWhnf` and adding another function `evaluateNf x = evaluateWhnf x >>= quote` (I'm hesitant to change `quote` to `evaluateNf` because I would expect all evalaution functions to start from the same point - an unevaluated `Term`)